src/CVRF/cvrf_priv.c

```diff
@@ -0,0 +1,2367 @@
+#define TAG_URL BAD_CAST "URL"
+#define TAG_VERSION BAD_CAST "Version"
+
+/*---------------------------------------------------------------------------------*\
+|							CVRF Parsing Helper Functions							|
+\*---------------------------------------------------------------------------------*/
+
+static void cvrf_set_parsing_error(const char *element) {
+	oscap_seterr(OSCAP_EFAMILY_XML, "Could not parse CVRF file: Missing or invalid"
+		"%s element\n", element);
+}
+
+static void cvrf_parse_container(xmlTextReaderPtr reader, struct oscap_list *list) {
+	cvrf_item_type_t item_type = cvrf_item_type_from_text((char *)xmlTextReaderConstLocalName(reader));
+	const char *tag = cvrf_item_type_get_text(item_type);
+	if (item_type != CVRF_VULNERABILITY && item_type != CVRF_VULNERABILITY_CWE)
+		xmlTextReaderNextElement(reader);
+	bool error = false;
+	while (xmlStrcmp(xmlTextReaderConstLocalName(reader), BAD_CAST tag) == 0) {
+		if (item_type == CVRF_REVISION) {
+			error = !oscap_list_add(list, cvrf_revision_parse(reader));
+		} else if (item_type == CVRF_NOTE || item_type == CVRF_DOCUMENT_NOTE) {
+			error = !oscap_list_add(list, cvrf_note_parse(reader));
+		} else if (item_type == CVRF_REFERENCE || item_type == CVRF_DOCUMENT_REFERENCE) {
+			error = !oscap_list_add(list, cvrf_reference_parse(reader));
+		} else if (item_type == CVRF_ACKNOWLEDGMENT) {
+			error = !oscap_list_add(list, cvrf_acknowledgment_parse(reader));
+		} else if (item_type == CVRF_GROUP) {
+			error = !oscap_list_add(list, cvrf_group_parse(reader));
+		} else if (item_type == CVRF_INVOLVEMENT) {
+			error = !oscap_list_add(list, cvrf_involvement_parse(reader));
+		} else if (item_type == CVRF_PRODUCT_STATUS) {
+			error = !oscap_list_add(list, cvrf_product_status_parse(reader));
+		} else if (item_type ==  CVRF_REMEDIATION) {
+			error = !oscap_list_add(list, cvrf_remediation_parse(reader));
+		} else if (item_type == CVRF_THREAT) {
+			error = !oscap_list_add(list, cvrf_threat_parse(reader));
+		} else if (item_type == CVRF_SCORE_SET) {
+			error = !oscap_list_add(list, cvrf_score_set_parse(reader));
+		} else if (item_type == CVRF_VULNERABILITY) {
+			error = !oscap_list_add(list, cvrf_vulnerability_parse(reader));
+		} else if (item_type == CVRF_VULNERABILITY_CWE) {
+			error = !oscap_list_add(list, cvrf_vulnerability_cwe_parse(reader));
+		}
+
+		xmlTextReaderNextNode(reader);
+		if (error) {
+			cvrf_set_parsing_error(tag);
+			break;
+		}
+	}
+}
+
+static char *cvrf_parse_element(xmlTextReaderPtr reader, const char *tagname, bool next_elm) {
+	char *elm_value = NULL;
+	if (xmlStrcmp(xmlTextReaderConstLocalName(reader), BAD_CAST tagname) == 0) {
+		elm_value = oscap_element_string_copy(reader);
+		if (next_elm)
+			xmlTextReaderNextElement(reader);
+	}
+	return elm_value;
+}
+
+static int cvrf_parse_ordinal(xmlTextReaderPtr reader) {
+	char *attribute = (char *)xmlTextReaderGetAttribute(reader, ATTR_ORDINAL);
+	int ordinal = strtol(attribute, NULL, 10);
+	free(attribute);
```

Comment:
Use xmlFree

Reply:
Well, this pointer could have been allocated by libxml2 or oscap_strdup. On linux, it doesn't make any difference, but in Windows we should free resources allocated by libxml2 with xmlFree.
Unfortunately, there is no way to distinguish it the allocator here.